### PR TITLE
Closure

### PR DESCRIPTION
**Pull request by ranjitjhala**
Super mega hacky and wobbly support for `Closure`. Specifically, this works:

```rust
#[flux::sig(fn(c: Option<bool>) -> Option<i32{v:10 <= v}>)]
pub fn test0(c: Option<bool>) -> Option<i32> {
    c.map(|b| if b { 1 } else { 2 }) //~ ERROR: postcondition
}

#[flux::sig(fn(c: Option<bool>) -> Option<i32{v:1 <= v}>)]
pub fn test1(c: Option<bool>) -> Option<i32> {
    c.map(|b| if b { 1 } else { 2 })
}
```

### CODE REVIEW

**Comment by ranjitjhala** [comment]
@nilehmann I think this is ready to look at now. We could iterate on this PR or merge and then you subject to whatever refactoring?

**Comment by ranjitjhala** [comment]
@nilehmann I think you can look at it now (or wait for the `build/check-tests` to pass...)

File: flux-driver/src/lib.rs [hunk 1]
```diff
@@ -14,7 +14,7 @@ extern crate rustc_span;
 
 mod callbacks;
 mod collector;
-mod mir_storage;
+// mod mir_storage;
```
Comment (nilehmann) [inline flux-driver/src/lib.rs:17]:
```suggestion
```

File: flux-middle/src/global_env.rs [hunk 2]
```diff
@@ -120,7 +120,7 @@ impl<'sess, 'tcx> GlobalEnv<'sess, 'tcx> {
             let variants = rty::conv::ConvCtxt::conv_struct_def_variant(self, struct_def)
                 .map(|v| vec![v.normalize(&self.defns)]);
             if config::dump_fhir() {
-                dbg::dump_item_info(self.tcx, def_id, "rty", &variants).unwrap();
+                dbg::dump_item_info(self.tcx, def_id.to_def_id(), "rty", &variants).unwrap();
```
Comment (nilehmann) [inline flux-middle/src/global_env.rs:123]:
```suggestion
                dbg::dump_item_info(self.tcx, def_id, "rty", &variants).unwrap();
```

File: flux-middle/src/global_env.rs [hunk 3]
```diff
@@ -148,7 +148,7 @@ impl<'sess, 'tcx> GlobalEnv<'sess, 'tcx> {
         for (def_id, fn_sig) in map.fn_sigs() {
             let fn_sig = rty::conv::conv_fn_sig(self, fn_sig).normalize(&self.defns);
             if config::dump_rty() {
-                dbg::dump_item_info(self.tcx, def_id, "rty", &fn_sig).unwrap();
+                dbg::dump_item_info(self.tcx, def_id.to_def_id(), "rty", &fn_sig).unwrap();
```
Comment (nilehmann) [inline flux-middle/src/global_env.rs:151]:
```suggestion
                dbg::dump_item_info(self.tcx, def_id, "rty", &fn_sig).unwrap();
```

File: flux-refineck/src/checker.rs [hunk 4]
```diff
@@ -78,16 +82,19 @@ pub(crate) trait Phase: Sized {
...
 
-    fn clear(&mut self, bb: BasicBlock);
+    fn clear(&mut self, def_id: DefId, bb: BasicBlock);
 }
 
-pub struct Inference<'a> {
-    bb_envs: &'a mut FxHashMap<BasicBlock, TypeEnvInfer>,
+pub struct Inference {
+    // bb_envs: &'a mut FxHashMap<BasicBlock, TypeEnvInfer>,
+    // bb_envs: &'a mut InferResult,
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:90]:
```suggestion
```

File: flux-refineck/src/checker.rs [hunk 5]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
     }
 }
 
-impl<'a, 'tcx> Checker<'a, 'tcx, Inference<'_>> {
+// type InferResult = FxHashMap<BasicBlock, TypeEnvInfer>;
+
+/// Inferred TypeEnv (after the INFER phase) for a single function (DefId)
+pub struct InferResult1 {
+    pub inner: FxHashMap<BasicBlock, TypeEnvInfer>,
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:140]:
```suggestion
    inner: FxHashMap<BasicBlock, TypeEnvInfer>,
```

File: flux-refineck/src/checker.rs [hunk 6]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
+}
+
+/// Inferred TypeEnv (after the INFER phase) for each "top-level" function + invoked closures
+pub struct InferResult {
+    pub inner: FxHashMap<DefId, InferResult1>,
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:145]:
```suggestion
    inner: FxHashMap<DefId, InferResult1>,
```

File: flux-refineck/src/checker.rs [hunk 7]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
     pub fn infer(
         genv: &GlobalEnv<'a, 'tcx>,
         body: &Body<'tcx>,
         def_id: DefId,
-    ) -> Result<FxHashMap<BasicBlock, TypeEnvInfer>, CheckerError> {
+    ) -> Result<InferResult, CheckerError> {
         dbg::infer_span!(genv.tcx, def_id).in_scope(|| {
-            let mut bb_envs = FxHashMap::default();
+            // let mut bb_envs = FxHashMap::default();
+            // let mut bb_envs = InferResult::new();
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:184]:
```suggestion
```

File: flux-refineck/src/checker.rs [hunk 8]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
+                span_bug!(body.span(), "checking function with unsupported signature")
+            });
+
             Checker::run(
                 genv,
                 RefineTree::new().as_subtree(),
-                body,
                 def_id,
-                Inference { bb_envs: &mut bb_envs },
+                &mut phase, // Inference { bb_envs: &mut bb_envs },
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:195]:
```suggestion
                &mut phase,
```

File: flux-refineck/src/checker.rs [hunk 9]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
+        bb_envs_infer: InferResult,
+    ) -> Result<KVarStore, CheckerError> {
+        let fn_sig = genv.lookup_fn_sig(def_id).unwrap_or_else(|_| {
+            span_bug!(body.span(), "checking function with unsupported signature")
+        });
+        let mut phase = Check { bb_envs_infer, bb_envs: FxHashMap::default(), kvars };
+
+        // TODO(CLOSURE): I broke this check_span thing, help!
+        dbg::check_span!(genv.tcx, def_id /* , bb_envs */)
+            .in_scope(|| Checker::run(genv, refine_tree, def_id, &mut phase, fn_sig))?;
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:220]: This breaks [this script](https://github.com/flux-rs/flux/blob/e3bfb5a257b26bfdb2185ec7bf2437168e433388/tools/logreader.py), but we can fix it later.

File: flux-refineck/src/checker.rs [hunk 10]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
+        // let rust_sig = genv.tcx.fn_sig(def_id);
+        // println!("TRACE: Checker::run (1) for {def_id:?} with RUST-SIG = {rust_sig:?}, POLY-SIG {fn_sig:?} and body:\n{body:?}");
+
+        // rustc_middle::mir::pretty::write_mir_fn(
+        //     genv.tcx,
+        //     bomir,
+        //     &mut |_, _| Ok(()),
+        //     &mut dbg::writer_for_item(self.genv.tcx, def_id.to_def_id(), "mir").unwrap(),
+        // )
+        // .unwrap();
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:249]:
```suggestion
```

File: flux-refineck/src/checker.rs [hunk 11]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
-        let env = Self::init(&mut rcx, body, &fn_sig);
+        let env = Self::init(&mut rcx, &body, &fn_sig);
 
         let dominators = body.dominators();
-        let mut ck = Checker::new(genv, def_id, body, fn_sig.output().clone(), &dominators, phase);
+
+        phase.enter_def_id(def_id);
+
+        let mut ck = Checker::new(genv, def_id, &body, fn_sig.output().clone(), &dominators, phase);
+        // OLD let mut ck = Checker::new(genv, def_id, body, fn_sig.output().clone(), &dominators, phase);
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:262]:
```suggestion
```

File: flux-refineck/src/checker.rs [hunk 12]
```diff
@@ -47,11 +47,12 @@ use crate::{
 };
 
 pub(crate) struct Checker<'a, 'tcx, P> {
+    pub def_id: DefId,
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:50]:
```suggestion
    def_id: DefId,
```

File: flux-refineck/src/checker.rs [hunk 13]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
     }
 }
 
-impl<'a, 'tcx> Checker<'a, 'tcx, Inference<'_>> {
+// type InferResult = FxHashMap<BasicBlock, TypeEnvInfer>;
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:136]:
```suggestion
```

File: flux-refineck/src/checker.rs [hunk 14]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
         genv: &'a GlobalEnv<'a, 'tcx>,
         mut refine_tree: RefineSubtree<'a>,
-        body: &'a Body<'tcx>,
         def_id: DefId,
-        phase: P,
+        phase: &'a mut P,
+        fn_sig: PolySig,
     ) -> Result<(), CheckerError> {
+        let body = {
+            let local_def_id = def_id.as_local().unwrap();
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:235]:
```suggestion
            let local_def_id = def_id.expect_local();
```

File: flux-refineck/src/checker.rs [hunk 15]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
     }
 }
 
-impl<'a, 'tcx> Checker<'a, 'tcx, Inference<'_>> {
+// type InferResult = FxHashMap<BasicBlock, TypeEnvInfer>;
+
+/// Inferred TypeEnv (after the INFER phase) for a single function (DefId)
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:138]:
```suggestion
/// Inferred [`TypeEnv`] (after the INFER phase) for a single function ([`DefId`])
```

File: flux-refineck/src/checker.rs [hunk 16]
```diff
@@ -125,67 +133,133 @@ impl<'a, 'tcx, P> Checker<'a, 'tcx, P> {
...
+pub struct InferResult1 {
+    pub inner: FxHashMap<BasicBlock, TypeEnvInfer>,
+}
+
+/// Inferred TypeEnv (after the INFER phase) for each "top-level" function + invoked closures
```
Comment (nilehmann) [inline flux-refineck/src/checker.rs:143]:
```suggestion
/// Inferred [`TypeEnv`] (after the INFER phase) for each "top-level" function + invoked closures
```

File: flux-refineck/src/constraint_gen.rs [hunk 17]
```diff
@@ -539,6 +551,150 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     fn solve(self) -> Result<EVarSol, UnsolvedEvar> {
         self.evar_gen.solve()
     }
+
+    fn mk_oblig(
```
Comment (nilehmann) [inline flux-refineck/src/constraint_gen.rs:555]: All the functions below only require the `GlobalEnv`, let's make them all free functions taking `genv: GlobalEnv<'_, 'tcx>` as the first parameter. You'll get an error if you try to call `closure_obligs` like so

```
        let closure_obligs = closure_obligs(self.genv, did, &substs, &actuals);
```

You can get around it by copying the the global env (`let genv = self.genv`) before calling `self.infcx`.

File: flux-refineck/src/constraint_gen.rs [hunk 18]
```diff
@@ -539,6 +551,150 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                self.mk_oblig(bound_vars, substs, oblig_def_id, &generics, p_rust_fn_sig)
+            })
+            .collect_vec()
+    }
+
+    fn gather_f_trait_info(
+        &mut self,
+        did: DefId,
+        rust_fn_sig: &rustc_middle::ty::PolyFnSig,
+        actuals: &[flux_middle::intern::Interned<rty::TyS>],
```
Comment (nilehmann) [inline flux-refineck/src/constraint_gen.rs:629]:
```suggestion
        actuals: &[rty::Ty],
```

File: flux-refineck/src/constraint_gen.rs [hunk 19]
```diff
@@ -539,6 +551,150 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                // found an fn-output type
+                Clause::Projection(proj_pred) => {
+                    let proj_ty = proj_pred.projection_ty;
+                    // ASSUME: the only trait item of `FnOnce` is `Output`
+                    if self.genv.tcx.trait_of_item(proj_ty.def_id) == Some(fn_once_id) &&
+                           let Some(proj_substs) = proj_ty.substs.try_as_type_list() &&
+                           let rustc_middle::ty::Param(p) = proj_substs[0].kind() &&
+                           let Some(out_ty) = proj_pred.term.ty() {
+                           f_out.insert(*p,out_ty);
+                        }
```
Comment (nilehmann) [inline flux-refineck/src/constraint_gen.rs:672]: `FnOnce::Output` is also a [lang item](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/lang_items/enum.LangItem.html#variant.FnOnceOutput). You can get it with `self.genv.tcx.lang_items().fn_once_output().unwrap()`

File: flux-tests/tests/neg/surface/closure00.rs [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+
+#[flux::sig(fn(c: Option<bool>) -> Option<i32{v:10 <= v}>)]
+pub fn test0(c: Option<bool>) -> Option<i32> {
+    c.map(|b| if b { 1 } else { 2 }) //~ ERROR: postcondition
+}
+
+#[flux::sig(fn(c: Option<bool>) -> Option<i32{v:1 <= v}>)]
+pub fn test1(c: Option<bool>) -> Option<i32> {
+    c.map(|b| if b { 1 } else { 2 })
+}
```
Comment (nilehmann) [inline flux-tests/tests/neg/surface/closure00.rs:12]: This should go in `pos/surface`.

Maybe we can also add a slightly more interesting test like

```
#[flux::sig(fn(c: Option<bool[true]>) -> Option<i32[1]>)]
pub fn test1(c: Option<bool>) -> Option<i32> {
    c.map(|b| if b { 1 } else { 2 })
}
```

**Review by nilehmann — APPROVED** [review]
@ranjitjhala Mostly cosmetic suggestions. The most important suggestions are for `constr_gen.rs`

File: flux-refineck/src/constraint_gen.rs (hunk 18 above)
Comment (ranjitjhala) [reply flux-refineck/src/constraint_gen.rs:629]: done!

File: flux-refineck/src/constraint_gen.rs (hunk 19 above)
Comment (ranjitjhala) [reply flux-refineck/src/constraint_gen.rs:672]: good catch!

File: flux-tests/tests/neg/surface/closure00.rs (hunk 20 above)
Comment (ranjitjhala) [reply flux-tests/tests/neg/surface/closure00.rs:12]: done!

File: flux-refineck/src/constraint_gen.rs (hunk 17 above)
Comment (ranjitjhala) [reply flux-refineck/src/constraint_gen.rs:555]: done.

**Pull request merged**